text/0000-convert-id.md

```diff
@@ -0,0 +1,204 @@
++ [Ruby](http://ruby-doc.org/core-2.5.0/Object.html#method-i-itself), which exports it to what amounts to the top type.
++ [Racket](http://docs.racket-lang.org/reference/values.html)
++ [Julia](https://docs.julialang.org/en/release-0.4/stdlib/base/#Base.identity)
++ [R](https://stat.ethz.ch/R-manual/R-devel/library/base/html/identity.html)
++ [F#](https://msdn.microsoft.com/en-us/visualfsharpdocs/conceptual/operators.id%5B%27t%5D-function-%5Bfsharp%5D)
++ [Clojure](https://clojuredocs.org/clojure.core/identity)
++ [Agda](http://www.cse.chalmers.se/~nad/repos/lib/src/Function.agda)
++ [Elm](http://package.elm-lang.org/packages/elm-lang/core/latest/Basics#identity)
+
+## The case for inclusion in the prelude
+
+Let's compare the effort required, assuming that each letter
+typed has a uniform cost wrt. effort.
+
+```rust
+use std::convert::identity; iter.filter_map(identity)
+
+fn identity<T>(x: T) -> T { x } iter.filter_map(identity)
+
+iter.filter_map(::std::convert::identity)
+
+iter.filter_map(identity)
+```
+
+Comparing the length of these lines, we see that there's not much difference in
+length when defining the function yourself or when importing or using an absolute
+path. But the prelude-using variant is considerably shorter. To encourage the
+use of the function, exporting to the prelude is therefore a good idea.
+
+In addition, there's an argument to be made from similarity to other things in
+`core::convert` as well as `drop` all of which are in the prelude. This is
+especially relevant in the case of `drop` which is also a trivial function.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+An identity function is a mapping of one type onto itself such that the output
+is the same as the input. In other words, a function `identity : T -> T` for
+some type `T` defined as `identity(x) = x`. This RFC adds such a function for
+all `Sized` types in Rust into libcore at the module `core::convert` and
+defines it as:
+
+```rust
+pub fn identity<T>(x: T) -> T { x }
+```
+
+This function is also re-exported to `std::convert::identity` as well as
+the prelude of both libcore and libstd.
+
+It is important to note that the input `x` passed to the function is
+moved since Rust uses move semantics by default.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+An identity function defined as `pub fn identity<T>(x: T) -> T { x }` exists as
+`core::convert::identity`. The function is also re-exported as
+`std::convert::identity` as well as the prelude of both libcore and libstd.
+
+Note that the identity function is not always equivalent to a closure
+such as `|x| x` since the closure may coerce `x` into a different type
+while the identity function never changes the type.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+It is already possible to do this in user code by:
+
++ using an identity closure: `|x| x`.
```

Comment:
What specific advantages do you see of `identity` over `|x| x`? `identity` is longer, and I think its meaning would be less obvious to most users, as they'd then have to go looking for the definition of a function called `identity`. I suppose there are potential codegen advantages, but those are small at best.

Reply:
A previous version of the RFC used `id`, but it was changed because `id` is ambiguous and for those who haven't heard of identity functions _very_ non obvious (it's easy to guess what it means if you see "identity", but not "id")

I assert that length is not the defining factor of ergonomics. `identity` is still easier to read and type over the closure, but I agree that the closure just easier to grok for those who haven't seen this function.

Reply:
> `identity` is still easier to read and type

I guess that's what I'm disputing, but it's totally a matter of opinion. I'd expect people familiar with Rust's syntax to quickly grok `|x| x`, while `identity` could require looking around for a function called `identity`.

Reply:
I can't speak for everyone, but I always double take when I see `|x| x` and the way I grok it is by going "oh right it's just the identity function"

Reply:
@cramertj 
> [..] I think its meaning would be less obvious to most users [..]

What users are you referring to specifically here? Current Rust users or all programmers in general? I think I believe you are correct in either case, but still - I have no evidence to offer myself that your assertion is correct, which is less than satisfactory.

> What specific advantages do you see of identity over |x| x? identity is longer [..]

Now that the length "advantage" (if we believe it is that) of `id` has been removed with the renaming to `identity`, I can only offer these thoughts:

+ A lot of functional languages and others have this function, so functional programmers may expect the function to exist while they may new to the language and don't understand what `|x| x` means without reading about it. Consistency with those (quite many) languages is an argument on its own.

+ Understanding `identity` for those that don't already understand (I think functional programmers will) it is a one time cost - this can also be said of closure syntax or any other identifier in the standard library, you have to learn it and that too is a one time setup cost. If you previously didn't know about `identity`, reading about it can also be a useful experience learning-wise.

+ I believe that using `identity` is more clearly showing that the identity-conversion was intentional compared to `|x| x`.

Reply:
@Centril Thanks for explaining! Those arguments seem reasonable to me. I personally still think that `identity` is less clear, but I'll leave it to others to voice their opinions one way or another, and I won't object to the feature if general consensus is that it would be beneficial.